File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums):
        pro=1
        array=[]
        num0=0
        for i in nums:
            if i!=0:
                pro*=i
            else:
                num0+=1
        
        for i in nums:
            if num0>=2:
                array.append(0)
            elif num0==0 :
                array.append(int(pro/i))
            else:
                if i==0: array.append(pro)
                else : array.append(0)

        return array

File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> resultat = new ArrayList<>();
        List<Integer> test = new ArrayList<>();
        
        backtrack(nums, target, 0, test, resultat);
        
        return resultat;
    }
    private void backtrack(int[] nums, int remaining, int start, List<Integer> test, List<List<Integer>> resultat) {
        if (remaining == 0) {
            resultat.add(new ArrayList<>(test));  // copie, car 'test' va continuer à être modifié
            return;
        }
        if (remaining < 0) {
            return;  // on a dépassé la cible, cette branche est invalide
        }
        
        for (int i = start; i < nums.length; i++) {
            test.add(nums[i]);
            // on rappelle avec 'i' (pas 'i+1') car on peut réutiliser le même nombre
            backtrack(nums, remaining - nums[i], i, test, resultat);
            test.remove(test.size() - 1);  // backtrack : on retire le dernier ajouté
        }
    }
}
